### PR TITLE
Add expected_logs

### PR DESCRIPTION
**Pull request by sophieyfang**
Add expected_logs and merge expected_logs and expected_metrics as one metadata.yaml file.

Details about the expected_logs are stated in the integration_test/README.md

### CODE REVIEW

File: integration_test/third_party_apps_test.go
Comment (jefferbrecht) [inline integration_test/third_party_apps_test.go:177]: Since this is still treated as a regex, a name like `Matcher` or `ValueRegex` might be more appropriate.

Comment (jefferbrecht) [reply integration_test/third_party_apps_test.go:177]: There's also a check in `constructQuery` for `Value != ""`, so you could actually make this optional as well. Or, for consistency with expected_metrics, we could default all `Value`s to `.*`. (The README says it's optional FWIW.)

File: integration_test/third_party_apps_test.go
Comment (jefferbrecht) [inline integration_test/third_party_apps_test.go:184]: These don't need to be pointers

File: integration_test/third_party_apps_test.go
Comment (jefferbrecht) [inline integration_test/third_party_apps_test.go:188]: Same, these don't need to be pointers

File: integration_test/third_party_apps_test.go
Comment (jefferbrecht) [inline integration_test/third_party_apps_test.go:193]: Update comment since it's not a map anymore

File: integration_test/third_party_apps_test.go
Comment (jefferbrecht) [inline integration_test/third_party_apps_test.go:227]:
```suggestion
		return fmt.Errorf("expected_metrics failed validation: %v", err)
```

File: integration_test/third_party_apps_test.go
Comment (jefferbrecht) [inline integration_test/third_party_apps_test.go:229]:
```suggestion
```

File: integration_test/third_party_apps_test.go
Comment (jefferbrecht) [inline integration_test/third_party_apps_test.go:412]:
```suggestion
			logger.ToMainLog().Println("found expected_metrics, running metrics test cases...")
```

File: integration_test/third_party_apps_test.go
Comment (jefferbrecht) [reply integration_test/third_party_apps_test.go:188]: Also, how does it behave if `expected_logs` or `expected_metrics` is missing? The nil check in `runSingleTest` suggests this should be allowed (i.e. they are optional) but there's no `validate:"optional"` here. Not sure if the marshaler will give an error or just give you an empty slice.

File: integration_test/README.md
Comment (jefferbrecht) [inline integration_test/README.md:181]:
```suggestion
We use `expected_logs` inside `metadata.yaml` both as a test artifact and as a source for documentation, e.g. [Apache(httpd) public docs](https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent/third-party/apache#monitored-logs). All logs ingested from the integration should be documented here.
```

File: integration_test/third_party_apps_test.go
Comment (jefferbrecht) [inline integration_test/third_party_apps_test.go:179]: `Description` is documented as required in the README.

File: integration_test/README.md
Comment (jefferbrecht) [inline integration_test/README.md:203]: s/`logEntry`/`LogEntry`

Also, since `value` behaves like a regex, that should be mentioned here.

File: integration_test/README.md
Comment (jefferbrecht) [inline integration_test/README.md:208]:
```suggestion
We use `expected_metrics` inside `metadata.yaml` both as a test artifact and as a source for documentation. All metrics ingested from the integration should be documented here.
```

**Comment by jefferbrecht** [comment]
This looks really good! Just some minor comments about docs/naming/optional/requiredness.

File: integration_test/third_party_apps_test.go
Comment (qingling128) [reply integration_test/third_party_apps_test.go:179]: `Description` should be required I guess? The public documentation is derived from this, right?

**Review by qingling128 — APPROVED** [review]
LGTM modulo the open threads get addressed (mainly minor comments so I'm gonna LGTM in advance)

File: integration_test/third_party_apps_test.go
Comment (sophieyfang) [reply integration_test/third_party_apps_test.go:188]: For activemq which only has metrics but no logs at this time, it will not fall under thiis code snippet so it will be no testing for logs
```
		if metadata.ExpectedMetrics != nil {
			logger.ToMainLog().Println("found expectedMetrics, running metrics test cases...")
			if err = runMetricsTestCases(ctx, logger, vm, metadata.ExpectedMetrics); err != nil {
				return nonRetryable, err
			}
		}
```
There is actually no built-in "optional" tag from golang validator: https://pkg.go.dev/github.com/go-playground/validator/v10. So i will remove the optional tag i put in else where

File: integration_test/third_party_apps_test.go
Comment (jefferbrecht) [reply integration_test/third_party_apps_test.go:179]: Yeah, it would be great IMO to keep it required. The validator should be changed from `optional` to `required` then.

File: integration_test/third_party_apps_test.go
Comment (sophieyfang) [reply integration_test/third_party_apps_test.go:177]: TL,DR: I will change Value to ValueRegex and make Value a required field

> [Optional. Not a very strong preference]
> 
> Making the field required could serve as a reminder for developers to verify the value in e2e whenever it is possible. It could potentially help boost test coverage. Developers can set it explicitly to `.*` when necessary.
> 
> This does mean the README needs to be changed to make it required though.
This is actually a very good point of view that i haven't thought of. I intended to make Value optional and description required but got it wrong..


> Since this is still treated as a regex, a name like `Matcher` or `ValueRegex` might be more appropriate.
`ValueRegex` SGTM

> There's also a check in `constructQuery` for `Value != ""`, so you could actually make this optional as well. Or, for consistency with expected_metrics, we could default all `Value`s to `.*`. (The README says it's optional FWIW.)
This is no build-in optional from https://pkg.go.dev/github.com/go-playground/validator/v10. For string, if the yaml file doesn't have it, it will default to "". If it's pointer, it will default to nil. Thinking that there will never be a case that logEntry field is intentionally set as "" and to be tested.

File: integration_test/third_party_apps_test.go
Comment (sophieyfang) [reply integration_test/third_party_apps_test.go:184]: hmmm initially i thought it might be more efficient to use pointer. codes in confgenerator seem to all use pointer. As the struct is not actually big one, maybe i am being over optimized.

File: integration_test/third_party_apps_test.go
Comment (sophieyfang) [reply integration_test/third_party_apps_test.go:193]: done

File: integration_test/README.md
Comment (sophieyfang) [reply integration_test/README.md:208]: done

File: integration_test/third_party_apps_test.go
Comment (jefferbrecht) [reply integration_test/third_party_apps_test.go]: How granular are the generated docs intended to be? I wonder if it would make sense to do per-leaf-field entries to solve the problem of objects, for example if something writes to `httpRequest` then have one entry for each field inside of `httpRequest`:

```yaml
- name: httpRequest.requestMethod
  value: GET|PUT|POST|...
- name: httpRequest.requestUrl
  value: .*
...
```

But that is a more complex question I think and we don't have to answer it here on the first cut.

I hadn't considered objects before so now I'm back to whether we should just make `ValueRegex` optional -- how about if the user doesn't specify it (or if they specify `""`), then instead of skipping the test, we check for field existence using e.g. `httpRequest:*`? This works for all kinds of fields, doesn't matter if it's object, string, int, etc.

File: integration_test/third_party_apps_data/applications/couchdb/metadata.yaml
Comment (jefferbrecht) [inline integration_test/third_party_apps_data/applications/couchdb/metadata.yaml:77]: Typo in `It's availble in access logs`, but I'm wondering what that part is trying to say. Is it saying that the information originated from couchdb's access logs? Or that the user can refer to the access logs for more information? I feel like the part before this sentence sufficiently explains the `httpRequest` anyway so maybe this whole chunk can be removed. Up to you.

File: integration_test/third_party_apps_data/applications/elasticsearch/metadata.yaml
Comment (jefferbrecht) [inline integration_test/third_party_apps_data/applications/elasticsearch/metadata.yaml:114]: These are intended to be literal `.` characters, so they need to be escaped.

File: integration_test/third_party_apps_test.go
Comment (jefferbrecht) [inline integration_test/third_party_apps_test.go:199]: I noticed that some of the `value_regex` contain spaces so the regex needs to be quoted:

```suggestion
			parts = append(parts, fmt.Sprintf(`%s=~"%q"`, field.Name, field.ValueRegex))
```

**Review by jefferbrecht — CHANGES_REQUESTED** [review]
I'm doing another pass and just leaving more minor/optional comments, but I did spot one thing we need to fix in the code, which is that the regex needs to be quoted.

File: integration_test/third_party_apps_data/applications/elasticsearch/metadata.yaml
Comment (jefferbrecht) [reply integration_test/third_party_apps_data/applications/elasticsearch/metadata.yaml:114]: (Since it was like that before, I'd be OK with doing a follow-up PR that does a deeper pass over the YAML files to resolve these kinds of issues. We don't need to hold up this PR for things that were already there.)

**Review by jefferbrecht — APPROVED** [review]
LGTM modulo a resolution to the `value_regex` thread 👍

**Pull request merged**
